File: first_unique_character_in_a_string.py

# https://leetcode.com/problems/first-unique-character-in-a-string/

class Solution:
	# O(n) time, O(n) space
	def firstUniqChar(self, s: str) -> int:
		char_index_repeated = {}
		for i in range(len(s)):
			c = s[i]
			if c not in char_index_repeated:
				char_index_repeated[c] = [i, False] 
			else:
				char_index_repeated[c][1] = True

		min_index = len(s)
		for key, val in char_index_repeated.items():
			if val[1] == False:
				min_index = min(min_index, val[0])

		return min_index if min_index != len(s) else -1

File: first_unique_character_in_a_string.cpp
#include <bits/stdc++.h>
using namespace std;

// https://leetcode.com/problems/first-unique-character-in-a-string/

class Solution {
public:
	// O(n) time.
	// O(1) space (27 lowercase letters).
	int firstUniqChar(string s) {
		int fst_unique_idx = -1;
		vector<int> char_freq(26);

		for(char c : s)
			++char_freq[c - 'a'];

		for(int i = 0; i < s.size(); ++i) {
			if(char_freq[s[i] - 'a'] == 1) {
				fst_unique_idx = i;
				break;
			}
		}

		return fst_unique_idx;
	}
};
